multihash.go

```diff
@@ -227,6 +229,36 @@ func FromB58String(s string) (m Multihash, err error) {
 	return Cast(b)
 }
 
+// FromBinary creates a new multihash from the binary
+// representation. The string is assumed to be a valid multihash.
+// This function will not alloc and will panic if it is unable to
+// parse the string.
+func FromBinary(v string) Multihash {
```

Comment:
Any reason not to have this return an error? I don't see this being called *so* often checking an error becomes onerous. For example, I expect we'll want to use this from go-cid when validating the CID's multihash.

(It may *also* be useful to have a `FromBinaryUnchecked` that doesn't even check but that's a separate issue).

Reply:
I wanted something that was fast and convenient to use when we _know_ the multihash is already valid.  The error checking that is done are sanity checks.  For example if the hash length is incorrect or bogus this will likely panic due to an out of bounds index check rather than return a proper error message.

I am somewhat against FromBinaryUnchecked if we can get away with (performance wise) always doing the checks.  This way we _know_ that the Multihash is valid if it is not empty.

See https://github.com/ipfs/go-cid/pull/73 for how I use this.  I am less against fixing this to return an error.

Reply:
Got it. I'd like to expose *some* `FromBinary` function that returns an error. Currently, `go-cid` will have to allocate to check the multihash. That is, it'll:

1. Pass a `[]byte` into `mh.Cast`.
2. `mh.Cast` will return a `string` backed multihash (allocating).

A `FromBinary` that returns an error would allow us to check without panicing or allocating.

Reply:
Okay.  I created such a function and called it `New`.

Reply:
Why `New`? It doesn't really create a *new* multihash.

> I am somewhat against FromBinaryUnchecked if we can get away with (performance wise) always doing the checks. This way we know that the Multihash is valid if it is not empty.

But we *aren't* always doing the checks! `FromString` leaves out one of the checks. Let's either check everything or nothing, checking somethings buys us nothing and just adds a bunch of code and two confusing functions that do slightly different things.

Reply:
Okay fair enough.  I renamed New to FromBinary when the Cid needs to return a Multihash it can just ignore the errors.

If necessary e can add an unchecked version if this becomes a performance problem (which I doubt).

Reply:
As you said, we can consider adding an unchecked variant later. My point was that we can actually put "checks" in the unchecked version (as long as they're fast enough) as long as we don't make any promises. My issue with the previous `FromBinary` incarnation is that it sort of checked and made unclear guarantees. I'd rather have a fully checked version (that returns the error) and a version that makes absolutely no guarantees (but *can* choose to check under the covers).

TL;DR: I'm happy with the current version, we can take a look into a less checked variant later.